File: 101._symmetric_tree/isSymmetric.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	if root != nil {
		return dfs(root.Left, root.Right)
	}

    return true
}

func dfs(left *TreeNode, right *TreeNode) bool {
	if left == nil || right == nil {
		return left == right
	}

	if left.Val != right.Val {
		return false
	}
	return dfs(left.Left, right.Right) && dfs(left.Right, right.Left)
}

File: 101._symmetric_tree/isSymmetric.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isSymmetric(root: TreeNode | null): boolean {
  if (!root) return true;

  return dfs(root.left, root.right);
}

function dfs(left: TreeNode | null, right: TreeNode | null): boolean {
  if (!left || !right) {
    return left == right;
  }

  if (left.val != right.val) return false;

  return dfs(left.left, right.right) && dfs(left.right, right.left);
}
